test/unit/object_origin.js

```diff
@@ -0,0 +1,351 @@
+    equal(rect.originX, 'left');
+
+    p = rect.adjustPosition('center');
+    deepEqual(rect.left, 55);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'center');
+
+    p = rect.adjustPosition('right');
+    deepEqual(rect.left, 75);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'right');
+
+    rect.originX = 'center';
+    rect.originY = 'center';
+    p = rect.adjustPosition('left');
+    deepEqual(rect.left, 55);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'left');
+
+    p = rect.adjustPosition('center');
+    deepEqual(rect.left, 75);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'center');
+
+    p = rect.adjustPosition('right');
+    deepEqual(rect.left, 95);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'right');
+
+    rect.originX = 'right';
+    rect.originY = 'bottom';
+    p = rect.adjustPosition('left');
+    deepEqual(rect.left, 55);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'left');
+
+    p = rect.adjustPosition('center');
+    deepEqual(rect.left, 75);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'center');
+
+    p = rect.adjustPosition('right');
+    deepEqual(rect.left, 95);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'right');
+  });
+
+  test('adjustPositionRotated', function(){
+    var rect = new fabric.Rect(rectOptions),
+        p,
+        point = new fabric.Point(15, 20);
+    rect.angle = 35;
+
+    rect.originX = 'left';
+    rect.originY = 'top';
+    p = rect.adjustPosition('left');
+    deepEqual(rect.left, 35);
+    deepEqual(rect.top, 45);
+    equal(rect.originX, 'left');
+
+    p = rect.adjustPosition('center');
+    deepEqual(rect.left, 51.383040885779835);
+    deepEqual(rect.top, 56.471528727020925);
+    equal(rect.originX, 'center');
+
+    p = rect.adjustPosition('right');
+    deepEqual(rect.left, 67.76608177155967);
+    deepEqual(rect.top, 67.94305745404185);
+    equal(rect.originX, 'right');
+
+    rect.originX = 'center';
+    rect.originY = 'center';
+    p = rect.adjustPosition('left');
+    deepEqual(rect.left, 51.383040885779835);
+    deepEqual(rect.top, 56.471528727020925);
+    equal(rect.originX, 'left');
+
+    p = rect.adjustPosition('center');
+    deepEqual(rect.left, 67.76608177155967);
+    deepEqual(rect.top, 67.94305745404185);
+    equal(rect.originX, 'center');
+
+    p = rect.adjustPosition('right');
+    deepEqual(rect.left, 84.1491226573395);
+    deepEqual(rect.top, 79.41458618106277);
+    equal(rect.originX, 'right');
+
+    rect.originX = 'right';
+    rect.originY = 'bottom';
+    p = rect.adjustPosition('left');
+    deepEqual(rect.left, 51.383040885779835);
+    deepEqual(rect.top, 56.47152872702093);
+    equal(rect.originX, 'left');
+
+    p = rect.adjustPosition('center');
+    deepEqual(rect.left, 67.76608177155967);
+    deepEqual(rect.top, 67.94305745404185);
+    equal(rect.originX, 'center');
+
+    p = rect.adjustPosition('right');
+    deepEqual(rect.left, 84.1491226573395);
```

Comment:
I'm a little unsure about such precise fractions. Usually they always cause problems. But maybe OK in this case.